pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -439,6 +437,23 @@ private boolean validate(TableConfig tableConfig, String taskType) {
     return true;
   }
 
+  /**
+   * Get the max valid merge bucket end time for the segment, LONG.MIN_VALUE if the segment is not ready to merge
+   */
+  private long getMaxValidBucketEndTimeMsForSegment(SegmentZKMetadata segmentZKMetadata, long bucketMs, long bufferMs) {
+    // Make sure the segment is ready to merge (the first bucket <= now - bufferTime)
+    if (segmentZKMetadata.getStartTimeMs() / bucketMs * bucketMs + bucketMs > System.currentTimeMillis() - bufferMs) {
+      return Long.MIN_VALUE;
+    }
+    // The max bucketEndTime is calculated as the min(segment end time, now - bufferTime) rounded to the bucket
+    // boundary. Notice bucketEndTime is exclusive while segment end time is inclusive.
+    // E.g. if bucketTime = 1d, the rounded segment end time of [10/1 00:00, 10/1 23:59] is 10/2 00:00
+    // The rounded segment end time of [10/1 00:00, 10/2 00:00] is 10/3 00:00
+    long maxBucketEndTimeMs = (segmentZKMetadata.getEndTimeMs() / bucketMs + 1) * bucketMs;
+    maxBucketEndTimeMs = Math.min(maxBucketEndTimeMs, (System.currentTimeMillis() - bufferMs) / bucketMs * bucketMs);
```

Comment:
Please store the value for `System.currentTimeMillis()` once and reuse the same value throughout the function.

Reply:
Updated.